### PR TITLE
.Net: Initial check-in of chat completion implementation

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

1. Non streaming chat completion support
2. Text embedding support

TODO
1. Streaming chat completion support
2. Unit test coverage
3. Integration tests

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs [hunk 1]
```diff
@@ -35,27 +40,123 @@ internal sealed class MistralClient
         this._logger = logger ?? NullLogger.Instance;
     }
 
-    internal Task<IReadOnlyList<ChatMessageContent>> GetChatMessageContentsAsync(ChatHistory chatHistory, PromptExecutionSettings? executionSettings = null, Kernel? kernel = null, CancellationToken cancellationToken = default)
+    internal async Task<IReadOnlyList<ChatMessageContent>> GetChatMessageContentsAsync(ChatHistory chatHistory, PromptExecutionSettings? executionSettings = null, Kernel? kernel = null, CancellationToken cancellationToken = default)
     {
-        throw new NotImplementedException();
+        string modelId = executionSettings?.ModelId ?? this._modelId;
+        var mistralExecutionSettings = MistralAIPromptExecutionSettings.FromExecutionSettings(executionSettings);
+        var request = this.CreateChatCompletionRequest(modelId, false, chatHistory, mistralExecutionSettings);
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs:47]: It looks like `boolean` parameter is passed as `false` here, and I don't see any other usages of `CreateChatCompletionRequest`. Is the goal to support streaming in the future, where this parameter will be passed as `true`?

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs [hunk 2]
```diff
@@ -35,27 +40,123 @@ internal sealed class MistralClient
...
+        var endpoint = new Uri($"{this._endpoint}{this._separator}chat/completions");
+
+        using var httpRequestMessage = this.CreatePost(request, endpoint, this._apiKey, false);
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs:50]: Let's use named parameters for better readability, especially for boolean values:
```suggestion
        using var httpRequestMessage = this.CreatePost(request, endpoint, this._apiKey, stream: false);
```

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs [hunk 3]
```diff
@@ -35,27 +40,123 @@ internal sealed class MistralClient
...
     #region private
     private readonly string _modelId;
-    private readonly string? _apiKey;
-    private readonly Uri? _endpoint;
+    private readonly string _apiKey;
+    private readonly Uri _endpoint;
     private readonly string _separator;
     private readonly HttpClient _httpClient;
     private readonly ILogger _logger;
+    private ChatCompletionRequest CreateChatCompletionRequest(string modelId, bool stream, ChatHistory chatHistory, MistralAIPromptExecutionSettings? executionSettings)
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs:80]: Small spacing:
```suggestion

    private ChatCompletionRequest CreateChatCompletionRequest(string modelId, bool stream, ChatHistory chatHistory, MistralAIPromptExecutionSettings? executionSettings)
```

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs [hunk 4]
```diff
@@ -35,27 +40,123 @@ internal sealed class MistralClient
...
+        request.Headers.Add(HttpHeaderConstant.Names.SemanticKernelVersion, HttpHeaderConstant.Values.GetAssemblyVersion(this.GetType()));
+        request.Headers.Add("Accept", stream ? "text/event-stream" : "application/json");
+        request.Headers.Add("Authorization", $"Bearer {this._apiKey}");
+        request.Content!.Headers.ContentType = new MediaTypeHeaderValue("application/json");
+    }
+
+    private async Task<T> SendRequestAsync<T>(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
+    {
+        //using var response = await this._httpClient.SendWithSuccessCheckAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+        using var response = await this._httpClient.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs:120]: Should we use `SendWithSuccessCheckAsync`? 
```suggestion
        using var response = await this._httpClient.SendWithSuccessCheckAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
```

File: dotnet/src/Connectors/Connectors.MistralAI/Client/MistralResponseBase.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Text.Json.Serialization;
+
+namespace Microsoft.SemanticKernel.Connectors.MistralAI.Client;
+
+/// <summary>
+///Base class for Mistral response.
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.MistralAI/Client/MistralResponseBase.cs:8]:
```suggestion
/// Base class for Mistral response.
```

File: dotnet/src/Connectors/Connectors.MistralAI/MistralChatMessageContent.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+using System.Collections.Generic;
+using Microsoft.SemanticKernel.ChatCompletion;
+using Microsoft.SemanticKernel.Connectors.MistralAI.Client;
+
+namespace Microsoft.SemanticKernel.Connectors.MistralAI;
+
+/// <summary>
+/// Mistral specialized chat message content
+/// </summary>
+public sealed class MistralChatMessageContent : ChatMessageContent
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.MistralAI/MistralChatMessageContent.cs:12]: Not sure I understand the purpose of this class, can we just return `ChatMessageContent`? It doesn't look like it has any differences with `ChatMessageContent` except naming.

File: dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        // Load configuration
+        this._configuration = new ConfigurationBuilder()
+            .AddJsonFile(path: "testsettings.json", optional: false, reloadOnChange: true)
+            .AddJsonFile(path: "testsettings.development.json", optional: true, reloadOnChange: true)
+            .AddEnvironmentVariables()
+            .AddUserSecrets<MistralAIChatCompletionTests>()
+            .Build();
+    }
+
+    [Fact] // (Skip = "This test is for manual verification.")
```
Comment (dmytrostruk) [inline dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs:29]: Should we remove this?
```suggestion
    [Fact]
```

File: dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.ChatCompletion;
+using Microsoft.SemanticKernel.Connectors.MistralAI;
+using Xunit;
+
+namespace SemanticKernel.IntegrationTests.Connectors.MistralAI;
+/// <summary>
```
Comment (dmytrostruk) [inline dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs:11]:
```suggestion

/// <summary>
```

File: dotnet/src/IntegrationTests/Connectors/MistralAI/TextEmbedding/MistralAITextEmbeddingTests.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using Microsoft.SemanticKernel.Connectors.MistralAI;
+using Xunit;
+
+namespace SemanticKernel.IntegrationTests.Connectors.MistralAI;
+/// <summary>
```
Comment (dmytrostruk) [inline dotnet/src/IntegrationTests/Connectors/MistralAI/TextEmbedding/MistralAITextEmbeddingTests.cs:10]:
```suggestion

/// <summary>
```

File: dotnet/src/IntegrationTests/Connectors/MistralAI/TextEmbedding/MistralAITextEmbeddingTests.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        // Load configuration
+        this._configuration = new ConfigurationBuilder()
+            .AddJsonFile(path: "testsettings.json", optional: false, reloadOnChange: true)
+            .AddJsonFile(path: "testsettings.development.json", optional: true, reloadOnChange: true)
+            .AddEnvironmentVariables()
+            .AddUserSecrets<MistralAITextEmbeddingTests>()
+            .Build();
+    }
+
+    [Fact] // (Skip = "This test is for manual verification.")
```
Comment (dmytrostruk) [inline dotnet/src/IntegrationTests/Connectors/MistralAI/TextEmbedding/MistralAITextEmbeddingTests.cs:28]: Same here?
```suggestion
    [Fact]
```

File: dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs (hunk 7 above)
Comment (markwallace-microsoft) [reply dotnet/src/IntegrationTests/Connectors/MistralAI/ChatCompletion/MistralAIChatCompletionTests.cs]: Adding it back in

**Review by dmytrostruk — APPROVED** [review]

**Pull request merged**